### PR TITLE
Avoid throwing explcit IOExceptions.

### PR DESCRIPTION
**Pull request by ankitaj224**
Also, handle slow network connection issues by retrying and back off.

### CODE REVIEW

File: firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java [hunk 1]
```diff
@@ -444,7 +444,7 @@ private String readExistingIidOrCreateFid(PersistedInstallationEntry prefs) {
 
   /** Registers the created Fid with FIS servers and update the persisted state. */
   private PersistedInstallationEntry registerFidWithServer(PersistedInstallationEntry prefs)
-      throws IOException {
+      throws FirebaseInstallationsException, IOException {
```
Comment (andirayo) [inline firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:447]: Can we just throw 1 exception type here?
Or where happens the unification (of exception types) for users?

File: firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java [hunk 2]
```diff
@@ -527,7 +532,7 @@ private Void deleteFirebaseInstallationId() throws FirebaseInstallationsExceptio
             /*projectID= */ getProjectIdentifier(),
             /*refreshToken= */ entry.getRefreshToken());
 
-      } catch (FirebaseException exception) {
+      } catch (FirebaseException | IOException exception) {
```
Comment (andirayo) [inline firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java:535]: Why do we catch `FirebaseException` here?

a) Do we ever use any Exception type other than `FirebaseInstallationsException`?
If yes: why? If no: I recommend to check for `FirebaseInstallationsException` here.
b) Why do we catch `FirebaseInstallationsException`?
catching and rethrowing the same exception is an antipattern.
Can we not just leave the FirebaseInstallationsException as is?

File: firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallationsException.java [hunk 3]
```diff
@@ -29,6 +29,11 @@
      * Project.
      */
     BAD_CONFIG,
+    /**
+     * The service is currently unavailable. This is a most likely a transient condition and may be
+     * corrected by retrying with a backoff.
```
Comment (andirayo) [inline firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallationsException.java:34]: What "backoff" are you referring to?
Exponential backoff?

If this is a public comment, I think I would either omit this addition or frame it like this:
"This is a most likely due to a transient condition and may be corrected by retrying. We recommend exponential backoff when retrying requests."

File: firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java [hunk 4]
```diff
@@ -138,7 +138,7 @@ public InstallationResponse createFirebaseInstallation(
       @NonNull String projectID,
       @NonNull String appId,
       @Nullable String iidToken)
-      throws IOException {
+      throws FirebaseInstallationsException, IOException {
```
Comment (andirayo) [inline firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java:141]: Why are we still throwing `IOException`?
Can we get around this and only throw 1 exception type?
See comment below.

File: firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java [hunk 5]
```diff
@@ -179,12 +179,16 @@ public InstallationResponse createFirebaseInstallation(
 
         // Return empty installation response with BAD_CONFIG response code after max retries
         return InstallationResponse.builder().setResponseCode(ResponseCode.BAD_CONFIG).build();
+      } catch (IOException ignored) {
+        retryCount++;
       } finally {
         httpURLConnection.disconnect();
```
Comment (andirayo) [inline firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java:185]: If we catch and ignore `IOException` here, can we remove `IOException` from the method interface?

File: firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java [hunk 6]
```diff
@@ -352,7 +360,7 @@ public TokenResult generateAuthToken(
       @NonNull String fid,
       @NonNull String projectID,
       @NonNull String refreshToken)
-      throws IOException {
+      throws FirebaseInstallationsException, IOException {
```
Comment (andirayo) [inline firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java:363]: Same question as before.
Btw, just as a comment, no action item for this PR: quite a bit of code duplication going on here.

File: firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java [hunk 7]
```diff
@@ -407,27 +419,37 @@ private static void logBadConfigError() {
...
-    if (heartbeatInfo != null && userAgentPublisher != null) {
-      HeartBeat heartbeat = heartbeatInfo.getHeartBeatCode(FIREBASE_INSTALLATIONS_ID_HEARTBEAT_TAG);
-      if (heartbeat != HeartBeat.NONE) {
-        httpURLConnection.addRequestProperty(USER_AGENT_HEADER, userAgentPublisher.getUserAgent());
-        httpURLConnection.addRequestProperty(
-            HEART_BEAT_HEADER, Integer.toString(heartbeat.getCode()));
+  private HttpURLConnection openHttpURLConnection(URL url, String apiKey)
+      throws FirebaseInstallationsException {
+    try {
+      HttpURLConnection httpURLConnection = (HttpURLConnection) url.openConnection();
```
Comment (andirayo) [inline firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java:425]: is it maybe enough to wrap this line with the try-catch block?
Maybe it gives more clarity about what can go wrong here.

Comment (ankitaj224) [reply firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java]: Done.

File: firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java (hunk 6 above)
Comment (ankitaj224) [reply firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java]: I will address the code duplication in another PR. Updated the exception handling.  PTAL. Thanks

File: firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java (hunk 2 above)
Comment (ankitaj224) [reply firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java]: Update to check only `FirebaseInstallationsException`.

Agreed on the anti-pattern. Updated the code to remove this code.

I don't know the reason why we had this anti-pattern behavior before.

File: firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java [hunk 8]
```diff
@@ -407,8 +412,16 @@ private static void logBadConfigError() {
             + " initializing Firebase.");
   }
 
-  private HttpURLConnection openHttpURLConnection(URL url, String apiKey) throws IOException {
-    HttpURLConnection httpURLConnection = (HttpURLConnection) url.openConnection();
+  private HttpURLConnection openHttpURLConnection(URL url, String apiKey)
+      throws FirebaseInstallationsException {
+    HttpURLConnection httpURLConnection;
+    try {
+      httpURLConnection = (HttpURLConnection) url.openConnection();
```
Comment (andirayo) [inline firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java:419]: Why are you casting this URLConnection?
Are you using methods that are specific to HttpURLConnection?
If not, I strongly recommend programming against the interface than the implementation.

File: firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java [hunk 9]
```diff
@@ -321,12 +313,27 @@ public void deleteFirebaseInstallation(
...
       }
     }
 
-    throw new IOException();
+    throw new FirebaseInstallationsException(
+        "Firebase Installations Service is unavailable. Please try again later.",
+        Status.UNAVAILABLE);
+  }
+
+  private URL getURL(String resourceName) throws FirebaseInstallationsException {
```
Comment (andirayo) [inline firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java:328]: +1 for adding this method!
An idea for a different name (feel free to ignore):
#getFullyQualifiedRequestUri
#getFullyQualifiedApiEndpoint

File: firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java [hunk 10]
```diff
@@ -321,12 +313,27 @@ public void deleteFirebaseInstallation(
...
+    try {
+      return new URL(
+          String.format(
+              "https://%s/%s/%s",
+              FIREBASE_INSTALLATIONS_API_DOMAIN, FIREBASE_INSTALLATIONS_API_VERSION, resourceName));
+    } catch (MalformedURLException e) {
+      throw new FirebaseInstallationsException(e.getMessage(), Status.UNAVAILABLE);
```
Comment (andirayo) [inline firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java:335]: This seems to be a pretty dramatic error.
How about we add some message here, e.g.
"Unexpected internal error while creating request URI."
I don't know.
Feel free to ignore.

File: firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java (hunk 8 above)
Comment (ankitaj224) [reply firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java:420]: Yes, I am using methods specific to HttpURLConnection

File: firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallationsException.java [hunk 11]
```diff
@@ -29,6 +29,12 @@
      * Project.
      */
     BAD_CONFIG,
+
+    /**
+     * The service is currently unavailable. This is a most likely due to a transient condition and
+     * may be corrected by retrying. We recommend exponential backoff when retrying requests."
```
Comment (ciarand) [inline firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallationsException.java:35]: extra quote?

**Pull request merged**